### PR TITLE
Add InferenceIds for sets theory.

### PR DESCRIPTION
**Pull request by nafur**
This PR introduces new InferenceId for the theory of sets and uses them instead of InferenceId::UNKNOWN.

### CODE REVIEW

File: src/theory/sets/inference_manager.cpp [hunk 1]
```diff
@@ -90,7 +90,7 @@ bool InferenceManager::assertFactRec(Node fact, Node exp, int inferType)
       || (atom.getKind() == EQUAL && atom[0].getType().isSet()))
   {
     // send to equality engine
-    if (assertInternalFact(atom, polarity, InferenceId::UNKNOWN, exp))
+    if (assertInternalFact(atom, polarity, InferenceId::SETS_EQUALITY, exp))
```
Comment (mudathirmahgoub) [inline src/theory/sets/inference_manager.cpp:93]: Shouldn't member and equal have different Ids here?

Comment (nafur) [reply src/theory/sets/inference_manager.cpp:93]: Possibly, should they just be `SETS_EQUALITY` and `SETS_MEMBER`?

Comment (ajreynol) [reply src/theory/sets/inference_manager.cpp:93]: Actually, this is another case where this method (assertFactRec) should take an `InferenceId`: this is a utility method used by many different inference schemas in sets.

**Comment by ajreynol** [comment]
I think something is off with the diff now?

**Comment by nafur** [comment]
> I think something is off with the diff now?

Hm, something was weird, I merged again. I added comments for some of the inferences already.

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/theory/sets/inference_manager.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/sets/inference_manager.cpp:93]: This one (lnie 93) still needs to change to id

File: src/theory/sets/inference_manager.cpp [hunk 2]
```diff
@@ -156,15 +160,15 @@ void InferenceManager::assertInference(std::vector<Node>& conc,
...
+  assertInference(conc, id, exp_n, c, inferType);
 }
 
 void InferenceManager::split(Node n, int reqPol)
 {
   n = Rewriter::rewrite(n);
   Node lem = NodeManager::currentNM()->mkNode(OR, n, n.negate());
   // send the lemma
-  lemma(lem, InferenceId::UNKNOWN);
+  lemma(lem, InferenceId::SETS_SPLIT);
```
Comment (ajreynol) [inline src/theory/sets/inference_manager.cpp:171]: Ditto, we should pass an `id` to this method.

File: src/theory/sets/inference_manager.h [hunk 3]
```diff
@@ -53,19 +53,26 @@ class InferenceManager : public InferenceManagerBuffered
    *
    * The argument c is the name of the inference, which is used for debugging.
    */
-  void assertInference(Node fact, Node exp, const char* c, int inferType = 0);
+  void assertInference(Node fact,
```
Comment (ajreynol) [inline src/theory/sets/inference_manager.h:56]: We can probably drop `c` from this method now, as it is subsumed by the id.  Similar to all of the below methods.

**Review by ajreynol — CHANGES_REQUESTED** [review]
Some more comments.

File: src/theory/sets/cardinality_extension.cpp [hunk 4]
```diff
@@ -418,7 +420,7 @@ void CardinalityExtension::checkCardCyclesRec(Node eqc,
           conc.push_back(n[e].eqNode(sib[e]));
         }
       }
-      d_im.assertInference(conc, n.eqNode(emp_set), "cg_emp");
+      d_im.assertInference(conc, InferenceId::UNKNOWN, n.eqNode(emp_set));
```
Comment (ajreynol) [inline src/theory/sets/cardinality_extension.cpp:423]:
```suggestion
      d_im.assertInference(conc, InferenceId::SETS_CARD_GRAPH_EMP, n.eqNode(emp_set));
```

File: src/theory/sets/cardinality_extension.cpp [hunk 5]
```diff
@@ -450,7 +452,8 @@ void CardinalityExtension::checkCardCyclesRec(Node eqc,
       {
         Trace("sets-debug") << "  it is empty..." << std::endl;
         Assert(!d_state.areEqual(n, emp_set));
-        d_im.assertInference(n.eqNode(emp_set), p.eqNode(emp_set), "cg_emppar");
+        d_im.assertInference(
+            n.eqNode(emp_set), InferenceId::UNKNOWN, p.eqNode(emp_set));
```
Comment (ajreynol) [inline src/theory/sets/cardinality_extension.cpp:456]:
```suggestion
            n.eqNode(emp_set), InferenceId::SETS_CARD_GRAPH_EMP_PARENT, p.eqNode(emp_set));
```

File: src/theory/sets/cardinality_extension.cpp [hunk 6]
```diff
@@ -497,7 +500,7 @@ void CardinalityExtension::checkCardCyclesRec(Node eqc,
         }
         Trace("sets-debug")
             << "...derived " << conc.size() << " conclusions" << std::endl;
-        d_im.assertInference(conc, n.eqNode(p), "cg_eqpar");
+        d_im.assertInference(conc, InferenceId::UNKNOWN, n.eqNode(p));
```
Comment (ajreynol) [inline src/theory/sets/cardinality_extension.cpp:503]:
```suggestion
        d_im.assertInference(conc, InferenceId::SETS_CARD_GRAPH_EQ_PARENT, n.eqNode(p));
```

File: src/theory/sets/cardinality_extension.cpp [hunk 7]
```diff
@@ -549,14 +552,14 @@ void CardinalityExtension::checkCardCyclesRec(Node eqc,
         if (eq_parent)
         {
           Node conc = n.eqNode(cpk);
-          d_im.assertInference(conc, exps, "cg_par_sing");
+          d_im.assertInference(conc, InferenceId::UNKNOWN, exps);
```
Comment (ajreynol) [inline src/theory/sets/cardinality_extension.cpp:555]:
```suggestion
          d_im.assertInference(conc, InferenceId::SETS_CARD_GRAPH_PARENT_SINGLETON, exps);
```

File: src/theory/sets/cardinality_extension.cpp [hunk 8]
```diff
@@ -604,7 +607,7 @@ void CardinalityExtension::checkCardCyclesRec(Node eqc,
               conc.push_back(cpk.eqNode(n));
             }
           }
-          d_im.assertInference(conc, cpk.eqNode(cpnl), "cg_pareq");
+          d_im.assertInference(conc, InferenceId::UNKNOWN, cpk.eqNode(cpnl));
```
Comment (ajreynol) [inline src/theory/sets/cardinality_extension.cpp:610]:
```suggestion
          d_im.assertInference(conc, InferenceId::SETS_CARD_GRAPH_EQ_PARENT_2, cpk.eqNode(cpnl));
```

File: src/theory/sets/theory_sets_private.cpp [hunk 9]
```diff
@@ -449,7 +449,7 @@ void TheorySetsPrivate::checkDownwardsClosure()
                 std::vector<Node> exp;
                 exp.push_back(mem);
                 exp.push_back(mem[1].eqNode(eq_set));
-                d_im.assertInference(nmem, exp, "downc");
+                d_im.assertInference(nmem, InferenceId::UNKNOWN, exp);
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_private.cpp:452]:
```suggestion
                d_im.assertInference(nmem, InferenceId::SETS_DOWN_CLOSURE, exp);
```

File: src/theory/sets/theory_sets_private.cpp [hunk 10]
```diff
@@ -474,7 +474,7 @@ void TheorySetsPrivate::checkDownwardsClosure()
                   nmem = NodeManager::currentNM()->mkNode(
                       kind::OR, pmem.negate(), nmem);
                 }
-                d_im.assertInference(nmem, exp, "downc");
+                d_im.assertInference(nmem, InferenceId::UNKNOWN, exp);
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_private.cpp:477]:
```suggestion
                d_im.assertInference(nmem, InferenceId::SETS_DOWN_CLOSURE, exp);
```

File: src/theory/sets/theory_sets_private.cpp [hunk 11]
```diff
@@ -578,7 +578,8 @@ void TheorySetsPrivate::checkUpwardsClosure()
                   {
                     Node kk = d_treg.getProxy(term);
                     Node fact = nm->mkNode(kind::MEMBER, x, kk);
-                    d_im.assertInference(fact, exp, "upc", inferType);
+                    d_im.assertInference(
+                        fact, InferenceId::SETS_MEMBER, exp, inferType);
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_private.cpp:582]:
```suggestion
                        fact, InferenceId::SETS_UP_CLOSURE, exp, inferType);
```

File: src/theory/sets/theory_sets_private.cpp [hunk 12]
```diff
@@ -605,7 +606,7 @@ void TheorySetsPrivate::checkUpwardsClosure()
                     d_state.addEqualityToExp(term[1], itm2m.second[1], exp);
                     Node r = d_treg.getProxy(term);
                     Node fact = nm->mkNode(kind::MEMBER, x, r);
-                    d_im.assertInference(fact, exp, "upc2");
+                    d_im.assertInference(fact, InferenceId::UNKNOWN, exp);
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_private.cpp:609]:
```suggestion
                    d_im.assertInference(fact, InferenceId::SETS_UP_CLOSURE_2, exp);
```

File: src/theory/sets/theory_sets_private.cpp [hunk 13]
```diff
@@ -667,7 +668,7 @@ void TheorySetsPrivate::checkUpwardsClosure()
                 exp.push_back(v.eqNode(it2.second[1]));
               }
               Node fact = nm->mkNode(kind::MEMBER, it2.second[0], u);
-              d_im.assertInference(fact, exp, "upuniv");
+              d_im.assertInference(fact, InferenceId::UNKNOWN, exp);
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_private.cpp:671]:
```suggestion
              d_im.assertInference(fact, InferenceId::SETS_UP_UNIV, exp);
```

File: src/theory/sets/theory_sets_private.cpp [hunk 14]
```diff
@@ -724,7 +725,7 @@ void TheorySetsPrivate::checkDisequalities()
     Node mem2 = nm->mkNode(MEMBER, x, deq[1]);
     Node lem = nm->mkNode(OR, deq, nm->mkNode(EQUAL, mem1, mem2).negate());
     lem = Rewriter::rewrite(lem);
-    d_im.assertInference(lem, d_true, "diseq", 1);
+    d_im.assertInference(lem, InferenceId::UNKNOWN, d_true, 1);
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_private.cpp:728]:
```suggestion
    d_im.assertInference(lem, InferenceId::SETS_DEQ, d_true, 1);
```

File: src/theory/sets/theory_sets_rels.cpp [hunk 15]
```diff
@@ -404,7 +404,7 @@ void TheorySetsRels::check(Theory::Effort level)
     if( distinct_skolems.size() >= 2 ) {
       conclusion =  NodeManager::currentNM()->mkNode( kind::AND, conclusion, NodeManager::currentNM()->mkNode( kind::DISTINCT, distinct_skolems ) );
     }
-    sendInfer(conclusion, reason, "JOIN-IMAGE DOWN");
+    sendInfer(conclusion, InferenceId::UNKNOWN, reason);
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_rels.cpp:407]:
```suggestion
    sendInfer(conclusion, InferenceId::SETS_RELS_JOIN_IMAGE_DOWN, reason);
```

File: src/theory/sets/theory_sets_rels.cpp [hunk 16]
```diff
@@ -437,8 +437,8 @@ void TheorySetsRels::check(Theory::Effort level)
       reason = NodeManager::currentNM()->mkNode( kind::AND, reason, NodeManager::currentNM()->mkNode( kind::EQUAL, exp[1], iden_term ) );
     }
     sendInfer(nm->mkNode(AND, fact, nm->mkNode(EQUAL, fst_mem, snd_mem)),
-              reason,
-              "IDENTITY-DOWN");
+              InferenceId::UNKNOWN,
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_rels.cpp:440]:
```suggestion
              InferenceId::SETS_RELS_IDENTITY_DOWN,
```

File: src/theory/sets/theory_sets_rels.cpp [hunk 17]
```diff
@@ -469,7 +469,8 @@ void TheorySetsRels::check(Theory::Effort level)
       if( (*mem_rep_exp_it)[1] != iden_term_rel ) {
         reason = NodeManager::currentNM()->mkNode( kind::AND, reason, NodeManager::currentNM()->mkNode( kind::EQUAL, (*mem_rep_exp_it)[1], iden_term_rel ) );
       }
-      sendInfer(nm->mkNode(MEMBER, new_mem, iden_term), reason, "IDENTITY-UP");
+      sendInfer(
+          nm->mkNode(MEMBER, new_mem, iden_term), InferenceId::UNKNOWN, reason);
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_rels.cpp:473]:
```suggestion
          nm->mkNode(MEMBER, new_mem, iden_term), InferenceId::SETS_RELS_IDENTITY_UP, reason);
```

File: src/theory/sets/theory_sets_rels.cpp [hunk 18]
```diff
@@ -712,12 +713,12 @@ void TheorySetsRels::check(Theory::Effort level)
     }
     if( all_reasons.size() > 1) {
       sendInfer(nm->mkNode(MEMBER, tc_mem, tc_rel),
-                nm->mkNode(AND, all_reasons),
-                "TCLOSURE-Forward");
+                InferenceId::UNKNOWN,
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_rels.cpp:716]:
```suggestion
                InferenceId::SETS_RELS_TCLOSURE_FWD,
```

File: src/theory/sets/theory_sets_rels.cpp [hunk 19]
```diff
@@ -712,12 +713,12 @@ void TheorySetsRels::check(Theory::Effort level)
...
+                nm->mkNode(AND, all_reasons));
     } else {
       sendInfer(nm->mkNode(MEMBER, tc_mem, tc_rel),
-                all_reasons.front(),
-                "TCLOSURE-Forward");
+                InferenceId::UNKNOWN,
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_rels.cpp:720]:
```suggestion
                InferenceId::SETS_RELS_TCLOSURE_FWD,
```

File: src/theory/sets/theory_sets_rels.cpp [hunk 20]
```diff
@@ -790,8 +791,8 @@ void TheorySetsRels::check(Theory::Effort level)
     if( pt_rel != exp[1] ) {
       reason = NodeManager::currentNM()->mkNode(kind::AND, exp, NodeManager::currentNM()->mkNode(kind::EQUAL, pt_rel, exp[1]));
     }
-    sendInfer(fact_1, reason, "product-split");
-    sendInfer(fact_2, reason, "product-split");
+    sendInfer(fact_1, InferenceId::UNKNOWN, reason);
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_rels.cpp:794]:
```suggestion
    sendInfer(fact_1, InferenceId::SETS_RELS_PRODUCT_SPLIT, reason);
```

File: src/theory/sets/theory_sets_rels.cpp [hunk 21]
```diff
@@ -790,8 +791,8 @@ void TheorySetsRels::check(Theory::Effort level)
...
+    sendInfer(fact_2, InferenceId::UNKNOWN, reason);
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_rels.cpp:795]:
```suggestion
    sendInfer(fact_2, InferenceId::SETS_RELS_PRODUCT_SPLIT, reason);
```

File: src/theory/sets/theory_sets_rels.cpp [hunk 22]
```diff
@@ -861,9 +862,9 @@ void TheorySetsRels::check(Theory::Effort level)
       reason = NodeManager::currentNM()->mkNode(kind::AND, reason, NodeManager::currentNM()->mkNode(kind::EQUAL, join_rel, exp[1]));
     }
     Node fact = NodeManager::currentNM()->mkNode(kind::MEMBER, mem1, join_rel[0]);
-    sendInfer(fact, reason, "JOIN-Split-1");
+    sendInfer(fact, InferenceId::UNKNOWN, reason);
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_rels.cpp:865]:
```suggestion
    sendInfer(fact, InferenceId::SETS_RELS_JOIN_SPLIT_1, reason);
```

File: src/theory/sets/theory_sets_rels.cpp [hunk 23]
```diff
@@ -861,9 +862,9 @@ void TheorySetsRels::check(Theory::Effort level)
...
     fact = NodeManager::currentNM()->mkNode(kind::MEMBER, mem2, join_rel[1]);
-    sendInfer(fact, reason, "JOIN-Split-2");
+    sendInfer(fact, InferenceId::UNKNOWN, reason);
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_rels.cpp:867]:
```suggestion
    sendInfer(fact, InferenceId::SETS_RELS_JOIN_SPLIT_2, reason);
```

File: src/theory/sets/theory_sets_rels.cpp [hunk 24]
```diff
@@ -885,8 +886,8 @@ void TheorySetsRels::check(Theory::Effort level)
     for( unsigned int i = 1; i < tp_terms.size(); i++ ) {
       Trace("rels-debug") << "\n[Theory::Rels] *********** Applying TRANSPOSE-Equal rule on transposed term = " << tp_terms[0] << " and " << tp_terms[i] << std::endl;
       sendInfer(nm->mkNode(EQUAL, tp_terms[0][0], tp_terms[i][0]),
-                nm->mkNode(EQUAL, tp_terms[0], tp_terms[i]),
-                "TRANSPOSE-Equal");
+                InferenceId::UNKNOWN,
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_rels.cpp:889]:
```suggestion
                InferenceId::SETS_RELS_TRANSPOSE_EQ,
```

File: src/theory/sets/theory_sets_rels.cpp [hunk 25]
```diff
@@ -911,8 +912,8 @@ void TheorySetsRels::check(Theory::Effort level)
       reason = NodeManager::currentNM()->mkNode(kind::AND, reason, NodeManager::currentNM()->mkNode(kind::EQUAL, tp_rel, exp[1]));
     }
     sendInfer(nm->mkNode(MEMBER, reversed_mem, tp_rel[0]),
-              reason,
-              "TRANSPOSE-Reverse");
+              InferenceId::UNKNOWN,
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_rels.cpp:915]:
```suggestion
              InferenceId::SETS_RELS_TRANSPOSE_REV,
```

File: src/theory/sets/theory_sets_rels.cpp [hunk 26]
```diff
@@ -1003,8 +1004,8 @@ void TheorySetsRels::check(Theory::Effort level)
           reason = NodeManager::currentNM()->mkNode(kind::AND, reason, NodeManager::currentNM()->mkNode(kind::EQUAL, rel[0], exps[i][1]));
         }
         sendInfer(nm->mkNode(MEMBER, RelsUtils::reverseTuple(exps[i][0]), rel),
-                  reason,
-                  "TRANSPOSE-reverse");
+                  InferenceId::UNKNOWN,
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_rels.cpp:1007]:
```suggestion
                  InferenceId::SETS_RELS_TRANSPOSE_REV,
```

File: src/theory/sets/theory_sets_rels.cpp [hunk 27]
```diff
@@ -1076,14 +1077,14 @@ void TheorySetsRels::check(Theory::Effort level)
             reasons.push_back( NodeManager::currentNM()->mkNode(kind::EQUAL, r2, r2_rep_exps[j][1]) );
           }
           if( isProduct ) {
-            sendInfer(fact,
-                      nm->mkNode(kind::AND, reasons),
-                      "PRODUCT-Compose");
+            sendInfer(
+                fact, InferenceId::UNKNOWN, nm->mkNode(kind::AND, reasons));
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_rels.cpp:1081]:
```suggestion
                fact, InferenceId::SETS_RELS_PRODUCE_COMPOSE, nm->mkNode(kind::AND, reasons));
```

File: src/theory/sets/theory_sets_rels.cpp [hunk 28]
```diff
@@ -1076,14 +1077,14 @@ void TheorySetsRels::check(Theory::Effort level)
...
           } else {
             if( r1_rmost != r2_lmost ) {
               reasons.push_back( NodeManager::currentNM()->mkNode(kind::EQUAL, r1_rmost, r2_lmost) );
             }
-            sendInfer(fact, nm->mkNode(kind::AND, reasons), "JOIN-Compose");
+            sendInfer(
+                fact, InferenceId::UNKNOWN, nm->mkNode(kind::AND, reasons));
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_rels.cpp:1087]:
```suggestion
                fact, InferenceId::SETS_RELS_JOIN_COMPOSE, nm->mkNode(kind::AND, reasons));
```

File: src/theory/sets/theory_sets_rels.cpp [hunk 29]
```diff
@@ -1235,7 +1236,7 @@ void TheorySetsRels::check(Theory::Effort level)
       Node tuple_reduct = NodeManager::currentNM()->mkNode(kind::APPLY_CONSTRUCTOR, tuple_elements);
       tuple_reduct = NodeManager::currentNM()->mkNode(kind::MEMBER,tuple_reduct, n[1]);
       Node tuple_reduction_lemma = NodeManager::currentNM()->mkNode(kind::EQUAL, n, tuple_reduct);
-      sendInfer(tuple_reduction_lemma, d_trueNode, "tuple-reduction");
+      sendInfer(tuple_reduction_lemma, InferenceId::UNKNOWN, d_trueNode);
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_rels.cpp:1239]:
```suggestion
      sendInfer(tuple_reduction_lemma, InferenceId::SETS_RELS_TUPLE_REDUCTION, d_trueNode);
```

**Review by ajreynol — CHANGES_REQUESTED** [review]
Let's make sure to add InferenceId for all the `c` that have meaningful information, sorry this is tedious.

File: src/theory/inference_id.h [hunk 30]
```diff
@@ -140,6 +140,57 @@ enum class InferenceId
   // enforces injectiveness of pto: (pto x y) ^ (pto y w) ^ x = y => y = w
   SEP_PTO_PROP,
 
+  // ***** sets theory
```
Comment (ajreynol) [inline src/theory/inference_id.h:143]:
```suggestion
  // ---------------------------------- sets theory
```

File: src/theory/inference_id.h [hunk 31]
```diff
@@ -140,6 +140,57 @@ enum class InferenceId
...
+  SETS_MEM_EQ,
+  SETS_MEM_EQ_CONFLICT,
+  SETS_PROXY,
+  SETS_PROXY_SINGLETON,
+  SETS_SINGLETON_EQ,
+  SETS_UP_CLOSURE,
+  SETS_UP_CLOSURE_2,
+  SETS_UP_UNIV,
+  SETS_UNIV_TYPE,
+
```
Comment (ajreynol) [inline src/theory/inference_id.h:158]: Let's avoid space between blocks belonging to the same theory.

File: src/theory/inference_id.h (hunk 30 above)
Comment (ajreynol) [reply src/theory/inference_id.h:143]: Also enclose the end.  Same goes for the Relations block below.

**Review by ajreynol — CHANGES_REQUESTED** [review]
I think you need to print the InferenceId now in inference_id.cpp.

**Review by ajreynol — APPROVED** [review]
LGTM, thanks for the all the updates.

**Pull request merged**